### PR TITLE
Strip unnecessary data from radosgw binary.

### PR DESCRIPTION
**Pull request by votdev**
Stripping the binary and shared libraries will reduce the size of the container image about 20MiB.

References:
- https://github.com/Debian/debhelper/blob/master/dh_strip
- https://www.thegeekstuff.com/2012/09/strip-command-examples/

Fixes: https://github.com/aquarist-labs/s3gw-core/issues/54

Signed-off-by: Volker Theile <vtheile@suse.com>

### CODE REVIEW

**Comment by votdev** [comment]
Are there additional `strip` arguments that we can add?

File: build/build-radosgw.sh [hunk 1]
```diff
@@ -39,6 +39,15 @@ build_radosgw_binary() {
   ninja -j${NPROC} bin/radosgw
 }
 
-build_radosgw_binary
+strip_radosgw() {
+  echo "Stripping files ..."
+  strip --strip-debug --strip-unneeded \
+    --remove-section=.comment --remove-section=.note \
```
Comment (m-ildefons) [inline build/build-radosgw.sh:45]: There only seems to be one `.comment` section, but there are multiple `.note` sections. Match them with `--remove-section=.note.*`

**Review by m-ildefons — COMMENTED** [review]
In addition to `strip`, but we should tell the compiler to optimize for size too:
`CMAKE_BUILD_TYPE=MINSIZEREL` which will enable `-Os -DNDEBUG`

**Review by trociny — COMMENTED** [review]
So, how are we going to debug radosgw crashes if they happen?

In Ceph world, after a daemon crash you had the crash backtrace logged, and it would be enough in 90% cases to identify the issue (and in the rest cases you would need to ask for the coredump). With a stripped binary I suppose that logged backtrace will be useless.

It is not that I disagree to have this change. Just wanted to mention.

**Comment by trociny** [comment]
Actually, I see you don't strip the symbol table, so I suppose the logged backtraces would look good enough to troubleshoot (the same we have in production).

Still, may be want to strip conditionally (e.g. skip stripping if an env variable is set)? So one could build a debug image. Ideally we could ship both stripped and debug images.

**Review by giubacc — COMMENTED** [review]
The patch seems fine to me, but I would add this feature as optional, I mean: having symbols is useful if you have to debug the core after a crash.
I would rather offer 2 images in our registry: 1 debug and 1 release.

**Comment by 0xavi0** [comment]
I agree, the possibility to have an image with all the information will be helpful in case of crashes or unexpected behaviour. 
So for release I'm 100% fine with stripping, but also having the option to have the "debug" container available.

**Comment by votdev** [comment]
The build type can be customized now via the environment variable `CMAKE_BUILD_TYPE`.

File: build/build-radosgw.sh [hunk 2]
```diff
@@ -2,20 +2,28 @@
 
 set -e
 
+# https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html
+# Release: Your typical release build with no debugging information and full optimization.
+# MinSizeRel: A special Release build optimized for size rather than speed.
+# RelWithDebInfo: Same as Release, but with debugging information.
+# Debug: Usually a classic debug build including debugging information, no optimization etc.
+CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE:-"MinSizeRel"}
```
Comment (giubacc) [inline build/build-radosgw.sh:10]: I would instead set default as: `Debug`.
I suspect this is the fastest build option and would be the preferred when developing.
When we will have a CI, we probably build the official images with that, and in that case the tool will call the script with a Release option.
It's my opinion btw.

Comment (votdev) [reply build/build-radosgw.sh]: Makes sense.

**Pull request merged**